crates/hir-def/src/lib.rs

```diff
@@ -1508,3 +1509,79 @@ pub struct UnresolvedMacro {
 
 #[derive(Default, Debug, Eq, PartialEq, Clone, Copy)]
 pub struct SyntheticSyntax;
+
+// Feature: Do Not Complete Attribute
+// Crate authors can opt their type out of completions in some cases.
+// This is done with the `#[rust_analyzer::do_not_complete(...)]` attribute.
+//
+// All completeable things support `#[rust_analyzer::do_not_complete(flyimport)]`,
+// which causes the thing to get excluded from flyimport completion. It will still
+// be completed when in scope. This is analogous to the setting `rust-analyzer.completion.autoimport.exclude`
+// with `"type": "always"`.
+//
+// In addition, traits support two more modes: `#[rust_analyzer::do_not_complete(methods_flyimport)]`,
+// which means the trait itself may still be flyimported but its methods won't, and
+// `#[rust_analyzer::do_not_complete(methods)]`, which means the methods won't be completed even when
+// the trait is in scope (but the trait itself may still be completed). The methods will still be completed
+// on `dyn Trait`, `impl Trait` or where the trait is specified in bounds. These modes correspond to
+// the settings `rust-analyzer.completion.autoimport.exclude` with `"type": "methods"` and
+// `rust-analyzer.completion.excludeTraits`, respectively.
+//
+// Malformed attributes will be ignored without warnings.
+//
+// Note that users have no way to override this attribute, so be careful and only include things
+// users definitely do not want to be completed!
+
+/// `#[rust_analyzer::do_not_complete(...)]` options.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub enum DoNotComplete {
+    /// No `#[rust_analyzer::do_not_complete(...)]`.
+    Complete,
+    /// `#[rust_analyzer::do_not_complete(flyimport)]`.
+    Flyimport,
+    /// `#[rust_analyzer::do_not_complete(methods_flyimport)]` (on a trait only).
+    MethodsFlyimport,
+    /// `#[rust_analyzer::do_not_complete(methods)]` (on a trait only).
+    Methods,
+}
```

Comment:
Okay how about?
```rust
/// `#[rust_analyzer::completions(...)]` options.
#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum Complete {
    /// No `#[rust_analyzer::completions(...)]`.
    Yes,
    /// `#[rust_analyzer::completions(ignore_flyimport)]`.
    IgnoreFlyimport,
    /// `#[rust_analyzer::completions(flyimport_ignore_methods)]` (on a trait only).
    IgnoreFlyimportMethods,
    /// `#[rust_analyzer::completions(ignore_methods)]` (on a trait only).
    IgnoreMethods,
}
```
that is a bit more general for future `completions` modifiers that may be positive / modifying

Reply:
Edited to that. But are you sure it needs to be `flyimport_ignore_methods` and not `ignore_flyimport_methods`?

Reply:
Oh I didn't even realize that I mixed that up. The latter seems more consistent I suppose so probably better to go with that?

Reply:
Edited.